pallets/loans/src/types/cashflow.rs

```diff
@@ -0,0 +1,348 @@
+		end: NaiveDate,
+		reference_date: ReferenceDate,
+	) -> Result<Vec<NaiveDate>, DispatchError>
+	where
+		Balance: FixedPointOperand,
+		Rate: FixedPointNumber,
+	{
+		// TODO: once we implement a pay_down_schedule other than `None`,
+		// we will need to adjust the expected interest amounts based on the
+		// expected outstanding principal at the time of the interest payment
+		Ok(match reference_date {
+			ReferenceDate::Date(reference) => {
+				let reference_date = NaiveDateTime::from_timestamp_opt(reference as i64, 0)
+					.ok_or(DispatchError::Other("Invalid origination date"))?
+					.date();
+
+				DateRule::monthly(start)
+					.with_end(end)
+					.with_rolling_day(reference_date.day())
+					.unwrap()
+					.into_iter()
+					// There's no interest payment expected on the origination date
+					.skip(1)
+					.collect()
+			}
+			ReferenceDate::End => DateRule::monthly(start)
+				.with_end(end)
+				.with_rolling_day(31)
+				.unwrap()
+				.into_iter()
+				.collect(),
+		})
+	}
+
+	fn add_interest_amounts<Balance, Rate>(
+		cashflows: Vec<NaiveDate>,
+		principal: Balance,
+		interest_payments: &InterestPayments,
+		interest_rate: &InterestRate<Rate>,
+		origination_date: Seconds,
+		maturity_date: Seconds,
+	) -> Result<Vec<(NaiveDate, Balance)>, DispatchError>
+	where
+		Balance: FixedPointOperand,
+		Rate: FixedPointNumber,
+	{
+		let cashflows_len = cashflows.len();
+		cashflows
+			.into_iter()
+			.enumerate()
+			.map(|(i, d)| -> Result<(NaiveDate, Balance), DispatchError> {
+				let interest_rate_per_sec = interest_rate.per_sec()?;
+				let amount_per_sec = interest_rate_per_sec.ensure_mul_int(principal)?;
+
+				if i == 0 {
+					// First cashflow: cashflow date - origination date * interest per day
+					let dt = d.and_hms_opt(0, 0, 0).ok_or("")?.timestamp() as u64;
+					return Ok((
+						d,
+						Rate::saturating_from_rational(
+							dt.ensure_sub(origination_date)?,
+							SECONDS_PER_YEAR,
+						)
+						.ensure_mul_int(amount_per_sec)?,
+					));
+				}
+
+				if i == cashflows_len {
+					//  Last cashflow: maturity date - cashflow date * interest per day
+					let dt = d.and_hms_opt(0, 0, 0).ok_or("")?.timestamp() as u64;
+					return Ok((
+						d,
+						Rate::saturating_from_rational(
+							maturity_date.ensure_sub(dt)?,
+							SECONDS_PER_YEAR,
+						)
+						.ensure_mul_int(amount_per_sec)?,
+					));
+				}
+
+				// Inbetween cashflows: interest per year / number of periods per year (e.g.
+				// divided by 12 for monthly interest payments)
+				let periods_per_year = match interest_payments {
+					InterestPayments::None => 0,
+					InterestPayments::Monthly(_) => 12,
+					InterestPayments::SemiAnnually(_) => 2,
+					InterestPayments::Annually(_) => 1,
+				};
+
+				let interest_rate_per_period = interest_rate
+					.per_year()
+					.ensure_div(Rate::saturating_from_integer(periods_per_year))?;
+				let amount_per_period = interest_rate_per_period.ensure_mul_int(principal)?;
```

Comment:
@hieronx I want to double check with you that this computation is correct.

This is performed for each element in the `cashflows` list, which is per month. But here, if the `interest_payments` is i.e. `Annually`, we are setting the whole year's interest to that cashflow month.

So, is this incorrect, or should I expect that if the user calls this with an `Annually` interest payment, then the user also sets a cash flow per year?